catalog/cataloger_merge.go

```diff
@@ -128,128 +165,99 @@ func hasCommitDifferences(tx db.Tx, leftID, rightID int64) (bool, error) {
-	nextCommitID, err := getNextCommitID(tx)
-	if err != nil {
-		return 0, err
+func mergeBatch(tx db.Tx, mergeBatch mergeBatchType, previousMaxCommitID, nextCommitID CommitID, rightID int64) error {
+	// collect changes to apply  on the branch
+	paths := make([]string, 0, MergeBatchSize)
+	ctidArray := make([]string, 0, MergeBatchSize)
+	var tombstonePaths []string
+	for _, diffRec := range mergeBatch {
+		if diffRec.DiffType == DifferenceTypeRemoved || diffRec.DiffType == DifferenceTypeChanged {
+			paths = append(paths, diffRec.Entry.Path)
+		}
+		if (diffRec.DiffType == DifferenceTypeAdded || diffRec.DiffType == DifferenceTypeChanged) &&
+			diffRec.EntryCtid != nil {
+			ctidArray = append(ctidArray, *diffRec.EntryCtid)
+		}
+		if diffRec.DiffType == DifferenceTypeRemoved && diffRec.parentTombstoneNeeded {
+			tombstonePaths = append(tombstonePaths, diffRec.Entry.Path)
+		}
 	}
-
-	// do the merge based on the relation
-	previousMaxCommitID, err := getLastCommitIDByBranchID(tx, rightID)
-	if err != nil {
-		return 0, err
+	// apply changes
+	if len(paths) > 0 {
+		// set entries that exist in the right branch as deleted by entries that were removed or changed
+		setMaxCommit := sq.Update("catalog_entries").
+			Set("max_commit", previousMaxCommitID).
+			Where("branch_id = ? and max_commit = ?", rightID, MaxCommitID).
+			Where(sq.Eq{"path": paths})
+		sql, args, err := setMaxCommit.PlaceholderFormat(sq.Dollar).ToSql()
+		if err != nil {
+			return err
+		}
+		_, err = tx.Exec(sql, args...)
+		if err != nil {
+			return err
+		}
 	}
-	switch relation {
-	case RelationTypeFromParent:
-		err = c.mergeFromParent(ctx, tx, previousMaxCommitID, nextCommitID, leftID, rightID, committer, msg, metadata)
-	case RelationTypeFromChild:
-		err = c.mergeFromChild(ctx, tx, previousMaxCommitID, nextCommitID, leftID, rightID, committer, msg, metadata)
-	case RelationTypeNotDirect:
-		err = c.mergeNonDirect(ctx, tx, previousMaxCommitID, nextCommitID, leftID, rightID, committer, msg, metadata)
-	default:
-		return 0, ErrUnsupportedRelation
+	if len(ctidArray) > 0 {
+		// copy entries from left to right
+		internalSelect := sq.Select(int64Str(rightID), "path", "physical_address", "creation_date", "size", "checksum", "metadata", int64Str(int64(nextCommitID))).
+			From("catalog_entries").
+			Where(sq.Eq{"ctid": ctidArray})
+		copyEntries := sq.Insert("catalog_entries").
+			Columns("branch_id", "path", "physical_address", "creation_date", "size", "checksum", "metadata", "min_commit").
+			Select(internalSelect)
+		sql, args, err := copyEntries.PlaceholderFormat(sq.Dollar).ToSql()
+		if err != nil {
+			return err
+		}
+		_, err = tx.Exec(sql, args...)
+		if err != nil {
+			return err
+		}
 	}
-	if err != nil {
-		return 0, err
+	// insert tombsotnes into parent branch that has a removed entry in its lineage
+	if len(tombstonePaths) > 0 {
+		values := "(VALUES ('" + strings.Join(tombstonePaths, "'),('") + "')) AS t(path)"
```

Comment:
Yes, That is certainly great. Tests I did when writing my blog post showed pgx array handling improved performance significantly

Reply:
## :warning: Given that path is user-controlled, this looks like a vector for an SQL injection attack vector by any user with permission to write a file to a branch.

Can you expand on this:
> Tried to use pgx array support, but it did not work for me

I discovered that it works in the export config stuff, e.g. https://github.com/treeverse/lakeFS/blob/29ae1cab03a8ef921b54696ad3a8179aa121d829/catalog/cataloger_export.go#L131 interpolates a `StringArray` type into an `INSERT`.
But failing that, we should wrap [`pgx.Connection.QuoteString`](https://pkg.go.dev/gopkg.in/jackc/pgx.v1#Connection.QuoteString) as part of `DB` to allow for safer quoting.  This version is unsafe (because of the quotes) and hard to read (because I need to wrap my head around the open and close brackets and quotes at each part of `strings.Join`.